File: algorithms/shell_sort.java
// Shell Short algorithm is an extension of and an improvement over the Insertion Sort.
// It uses a sequence of increments (gaps) to sort elements.
// It starts by sorting pairs of elements far apart from each other, then gradually reduces the gap between elements to perform a final Insertion Sort.
// The idea is to move elements faster towards their expected position.

// Imagine you have a list of tasks to do, each with a priority level.
// You want to sort these tasks based on priority but prefer to tackle similar priority tasks together for efficiency.

public class ShellSort {
    public static void shellSort(int[] arr) {
        int n = arr.length;

        // Start with a large gap, then reduce the gap
        for (int gap = n / 2; gap > 0; gap /= 2) {
            // Do a gapped insertion sort for this gap size
            for (int i = gap; i < n; i++) {
                int temp = arr[i];

                // Shift earlier gap-sorted elements up until the correct location
                int j;
                for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                    arr[j] = arr[j - gap];
                }

                // Place temp (the original arr[i]) in its correct location
                arr[j] = temp;
            }
        }
    }

}

File: algorithms/shell_sort.kt
// Shell Sort is a generalization of insertion sort that allows the exchange of items that are far apart.
// The idea is to arrange the list of elements so that, starting anywhere, taking every hth element produces a sorted list.
// Such a list is said to be h-sorted. It start by sorting pairs of elements far apart from each other, then progressively reducing the gap between elements to be compared.

// Imagine you're organizing a large set of library books by their ID numbers.
// Instead of comparing each book only with its immediate neighbor (as in insertion sort), you start by comparing books that are farther apart
// and gradually reduce this gap. This method can quickly bring books that are far out of place closer to their correct position, thus speeding up the sorting process.
// This is eseentially what the Shell Sort algorithm does.

fun shellSort(arr: IntArray) {
    var n = arr.size
    var gap = n / 2

    while (gap > 0) {
        for (i in gap until n) {
            val temp = arr[i]
            var j = i
            while (j >= gap && arr[j - gap] > temp) {
                arr[j] = arr[j - gap]
                j -= gap
            }
            arr[j] = temp
        }
        gap /= 2
    }
}
